### ISSUE TITLE
API point: Experiment matrix

### ISSUE THREAD

**Issue by nsheff** [databio/episb-hub#33]
Given a segmentation ID, return a (zipped) matrix with each row representing a segment, and each column representing an experiment that annotates that segmentation.

**Comment by oddodaoddo** [comment]
Off the top of my head, this will need to use scroll APIs from elastic.

**Comment by oddodaoddo** [comment]
Working on this now.

**Comment by oddodaoddo** [comment]
This [commit](https://github.com/databio/episb-provider/commit/3635b990a483931ec7ceb7ae990a43c9e22ffb4d) implements functionality on the provider to get a .gzipped file of an experiment (annotation/segmentID value rows). To get the file, simply indicate to the server that you can accept .gz content, such as:
```
curl -H "Accept-Encoding: gzip" http://provider.episb.org/episb-provider/experiments/get/BySegmentationName/broad?matrix=true --output some_file.gz
```

**Comment by oddodaoddo** [comment]
This uses the elastic scroll API to provide the FULL answer, which may be large.

**Comment by oddodaoddo** [comment]
@nsheff : Do we want some kind of a screen on episb.org that allows a person to enter a segmentation or is this more for programmatic use? Thanks!

**Comment by oddodaoddo** [comment]
Ping 😄

**Comment by nsheff** [comment]
at this point just programatic use is fine.

**Comment by oddodaoddo** [comment]
Ok - cool, thanks!

**Comment by oddodaoddo** [comment]
It appears I misunderstood the problem definition here. What we want as a reply from this query is a matrix of the following format, presented as BED file:

Segmentation::segmentID_1 Experiment_1 Experiment_2 .... Experiment_n
Segmentation::segmentID_2 AnnValue_1 AnnValue_2 ... AnnValue_n
.
.
.
Segmentation::segmentID_m AnnValue_1 AnnValue_2 ... AnnValue_n

@nsheff : is this correct?

If so, rolling this up in elastic may be a problem (may not be doable from elastic). This would mean that I would have to pull in all the data and roll it up myself. May be memory intensive. Another approach would be to deploy [Apache Ignite](http://ignite.apache.org) which would allow us to use SQL to query elastic and would make this much easier.

**Comment by oddodaoddo** [comment]
Elastic also has some SQL support so this may be an option as well.

**Comment by oddodaoddo** [comment]
Actually, the format requested would be:
chr start stop experiment_1 experiment_2 ... experiment_n
.
.

This would make it even more complicated since we only track annotations by their segment IDs, we do not store the chr/start/stop info with each annotation to avoid duplication of data.

**Comment by nsheff** [comment]
you would start with a segmentation, identify all segments in the segmentation, and then identify all experiments for each segment.

you already have an endpoint to identify the segments in the segmentation. And you have the endpoint to identify all experiments for each segment.

I'm a bit reluctant to introduce new technology at the early stage, rather wanting to first get things functional, even if not done in the most efficient way. this will not be a commonly executed query so it's fine if it takes a bit of processing.

**Comment by oddodaoddo** [comment]
The way we have the data structure in elastic and the APIs, I actually have an endpoint that retrieves all Annotations for a given Segmentation. Something like:
```
{"result":[{"segmentID":"DHS::b526f39e-b7ef-44bd-b959-31db63146802","annValue":0.20170000195503235,"experiment":{"experimentName":"CLL","experimentProtocol":"","experimentCellType":"","experimentSpecies":"Human","experimentTissue":"","experimentAntibody":"","experimentTreatment":"","experimentDescription":""},"study":{"studyAuthor":{"familyName":"Default","givenName":"Author","email":"info@episb.org"},"studyManuscript":"Default Manuscript","studyDescription":"","studyDate":""}},
```

However, this information does not have the chr/start/stop info, just the segment ID. In a real-life scenario, I would have to take the thousands and thousands of segment IDs and hit up elastic for each one to get the chr/start/stop info. The query would take hours ;)

**Comment by oddodaoddo** [comment]
Just to reiterate today's meeting: I will return a bed file of following format:

```
segmentation::segmentID_1 ann_value_1 ann_value_2 ... ann_value_n
segmentation::segmentID_2 ann_value_1 ann_value_2 ... ann_value_n
.
.
.
segmentation::segmentID_m ann_value_1 ann_value_2 ... ann_value_n
```

Correct?

**Comment by nsheff** [comment]
yes.

**Comment by oddodaoddo** [comment]
I have a fix deployed in the dev branch of episb-provider. The SQL route did not work from elastic so I resorted to manual labor. It is slow and I am working on making it faster. One problem is that due to the size of the data we need to use the scroll API from elastic - meaning, we are rolling up the data one window at a time. However, the size of this window needs to be carefully calculated so that we can get all the experiments for one segment ID at a time (we don't break them up between windows). Essentially, the algorithm is:

1) get all annotations for segmentation X, sorted by segment ID and then experiment name
2) make sure we received only the amount of rows that equals the number of experiments
3) process these so they end up in one line
4) write the line into a file
5) when done, serve the file

The problem is - we can have a 100 experiments and 3 million segments - meaning we are getting the data 100 lines at a time, each time hitting up elastic.

I am working on a fix where we are getting some multiple of number of experiments and processing the block instead. My head is burning...

**Comment by nsheff** [comment]
I have a new idea... Try this:

write an endpoint that searches for an experimentID and a segmentationID.
then:
1. get all experiments for a segmentation
2. Run a loop through all experiments/segmentation search
3. concat.

let me know if we can talk about it --

**Comment by oddodaoddo** [comment]
I will take that into consideration - it sounds like a simple idea and that's always good. I have sped up the query somewhat, it is live now on the server. For the DHS segmentation (which is pretty chunky), the file takes about 10-15 minutes to produce and pull down to my machine at home,which sounds pretty long.

**Comment by nsheff** [comment]
it' just changes the loop... you're looping over segments. it's better to loop over experiments.

**Comment by oddodaoddo** [comment]
Yeah... There are 3 million rows in DHS and 104 experiments - that makes for 3 million rows x 104 columns = a large matrix to produce from essentially 300 million json documents (one document for each annotation - segmentation linkage) pulled in from elastic - they all get parsed first to extract the fields, so on and so on...

**Comment by oddodaoddo** [comment]
Ok - here is what is happening: DHS segmentation has nearly 3 million segments. Each experiment has nearly 3 million annotations, meaning 3 million JSON documents saying "this annotation value is linked to DHS::segment_ID_something". 104 experiments linked to DHS segmentation means 104 x 3 million documents and roughly 300 million documents.

Elastic will let you get 10,000 documents each time via the scroll API (we can change this number to be higher at cost of memory). This means a lot of retrieval trips and then processing each chunk to roll it into the matrix.

That's where the problem is coming from. It would be nice if we could make elastic produce this matrix internally or at least each column of it (we would still have to get it via the scrolling API so lots of round trips). I am going to talk to a few friends of mine who know elastic in depth and see if they can offer any insights or anything I am missing.

**Comment by nsheff** [comment]
it seems to me that the ignite/sql solution is not going to solve this becuase it will still have to do these kinds of queries at some point.

**Comment by oddodaoddo** [comment]
Probably not, although Ignite is a distributed cache and being distributed and a cache, it may have some or a lot of the requested data already lying around. Ultimately, it would be nice if elastic can roll this up internally but horizontally scaling document/json "databases" are not really meant to be used as traditional rdbms systems. We could try a traditional RDBMS to store our data but scaling will be an issue at some point - I am thinking of scenarios where we have thousands of segmentations, each with millions of segments and thousands and thousands of experiments. Of course, I do not have enough domain knowledge to know whether this is a real expectation or not but as an engineer I have to imagine very large worst case scenarios.

**Comment by oddodaoddo** [comment]
I will say this though: the machine we are running on has 72 cores. If we could fire up 70 concurrent threads, all asking for their own elastic chunk and processing it to add to the matrix, I can see it finishing up pretty quick. We can raise the scroll window size to 50,000 and it could end up being fairly decent performance. This is only one machine - spreading the computation across a cluster would make it that much faster...
